File: Problems/programmers_다리를지나는트럭.py
#https://school.programmers.co.kr/learn/courses/30/lessons/42583
#DataStructure(Stack,Queue)

def solution(bridge_length, weight, truck_weights):
    b = [0]*bridge_length
    s = 0
    
    count = 0
    while b:
        count += 1
        s -= b.pop(0)
        
        if truck_weights:
            if s+truck_weights[0] <= weight:
                s += truck_weights[0]
                b.append(truck_weights.pop(0))
            else:
                b.append(0) 
    
    return count

#deque는 Queue로 쓰일때만 빠르다.
#stack으로 사용되는 경우는 list 써도 무방.

File: Problems/cpp/programmers_다리를지나는트럭.cpp
//https://school.programmers.co.kr/learn/courses/30/lessons/42583
//Implementation

#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer=0, tw=0;
    deque<int> bq(bridge_length, 0);
    deque<int> tq(truck_weights.begin(), truck_weights.end());
    
    while (!bq.empty()){
        answer++;
        tw -= bq.front();
        bq.pop_front();
        
        if (!tq.empty()){
            if (tw+tq.front() <= weight){
                tw += tq.front();
                bq.push_back(tq.front());
                tq.pop_front();
            }
            else
                bq.push_back(0);
        }
    }
    
    return answer;
}